### PR TITLE
Issue 73

### PR DESCRIPTION
**Pull request by JustinKyleJames**
Opening in draft mode while I figure out how to test this.

Ignore commit for issue 65 as this will be rebased on that once its pull request is finished.

### CODE REVIEW

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (korydraughn) [inline DSI/globus_gridftp_server_iRODS.cpp:2210]: `.data()` -> `.c_str()`

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (korydraughn) [inline DSI/globus_gridftp_server_iRODS.cpp:2012]: Is this needed? It doesn't appear to be used anywhere.

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (korydraughn) [inline DSI/globus_gridftp_server_iRODS.cpp:289]: How is `utime` set?
Is it guaranteed to be set to a valid value?

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (korydraughn) [inline DSI/globus_gridftp_server_iRODS.cpp:2210]: Log the iRODS error code?

**Review by korydraughn — COMMENTED** [review]
Seems good overall.

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (JustinKyleJames) [reply DSI/globus_gridftp_server_iRODS.cpp:2012]: No, not anymore.  I will remove it.

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (JustinKyleJames) [reply DSI/globus_gridftp_server_iRODS.cpp:289]: The address of it gets sent to the call to `globus_gridftp_server_get_recv_modification_time()`. That function will either set it to what the client requested or will set it to -1 if the client does not request it.  If that function returns an error, I set it to -1.

I only read that after calling `globus_gridftp_server_get_recv_modification_time()` so it will always be set properly.

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (JustinKyleJames) [reply DSI/globus_gridftp_server_iRODS.cpp:2210]: done

Comment (JustinKyleJames) [reply DSI/globus_gridftp_server_iRODS.cpp:2210]: I logged the numeric code.

**Comment by JustinKyleJames** [comment]
I got help from Globus on how to test this.  I added a test case for it.

**Comment by JustinKyleJames** [comment]
I rebased this on main after issue 65 was merged.

**Comment by trel** [comment]
looks good - this passes the tests?

are you going to squash to one?   or two (code+test)?

**Comment by JustinKyleJames** [comment]
> looks good - this passes the tests?
> 
> are you going to squash to one? or two (code+test)?

Yes, it passes the tests.  I'll squash to one I think.

**Comment by trel** [comment]
Seems good.

File: tests/globus_test.py
Comment (korydraughn) [inline tests/globus_test.py:220]: What is the unit of time for this value?

Does this value result in the expected timestamp appearing in iRODS?
It's probably a good idea to add an assertion for this using the icommands _(assuming the icommands are available here)_.

Comment (JustinKyleJames) [reply tests/globus_test.py:220]: This time is in format YYYYMMDDHHMISS.  That is just the way it is done.

I can't at the moment test it with icommands because I don't have the ability to run icommands as the same user as the anonymous FTP.  There is an enhancement open for adding the ability to run icommands for other users (similar to core tests).

Comment (korydraughn) [reply tests/globus_test.py:220]: Cool cool.

**Review by korydraughn — APPROVED** [review]
Looks good to me.

Let's get one more approval before pounding.

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (alanking) [inline DSI/globus_gridftp_server_iRODS.cpp:2053]: Is there any value in showing `result` in this error message?

**Review by alanking — COMMENTED** [review]
Just had one suggestion for a potential improvement. Otherwise, looks good.

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (JustinKyleJames) [reply DSI/globus_gridftp_server_iRODS.cpp]: I changed it to print the return code.  I am not sure what might be returned as there are only #defines for GLOBUS_SUCCESS (0) and GLOBUS_FAILURE (-1).

**Comment by JustinKyleJames** [comment]
I updated the commit message to state "preserve file modification time" instead of "preserve replica".  I think my brain was on autopilot when I wrote the original message.

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (korydraughn) [reply DSI/globus_gridftp_server_iRODS.cpp]: Consider asking the Globus team about whether other values can be returned.

If the only values ever returned are 0 or -1, then I don't see much value in logging it.

Thoughts?

File: tests/globus_test.py
Comment (trel) [reply tests/globus_test.py:220]: ISO8601 usually has a `T` between the date and the time part.   `YYYYMMDDThhmmss`

is this human visible/parsable/usable by others?  or just the test itself?

File: DSI/globus_gridftp_server_iRODS.cpp
Comment (JustinKyleJames) [reply DSI/globus_gridftp_server_iRODS.cpp]: > Consider asking the Globus team about whether other values can be returned.
> 
> If the only values ever returned are 0 or -1, then I don't see much value in logging it.
> 
> Thoughts?

I went ahead and logged the error the way the Globus guy suggested. 

```
globus_gfs_log_result(GLOBUS_GFS_LOG_WARN, "iRODS: Error getting modtime, skipping: ", result); 
```

Comment (JustinKyleJames) [reply DSI/globus_gridftp_server_iRODS.cpp]: They suggested warn and after seeing it I agreed that that is probably the better approach since this wouldn't be a fatal issue.

Here is the response I got back:

```
For a globus_result_t, anything other than GLOBUS_SUCCESS is error.

You can log the error:
{
     globus_gfs_log_result(GLOBUS_GFS_LOG_WARN, "Error getting modtime,
skipping:", result);
}

Or get the error string for your own logging:
{
     char *                              msg;
     msg = globus_error_print_friendly(globus_error_peek(result));
     ...
     globus_free(msg);
}

With some globus_result_t types, you can check the module-specific error
codes with additional globus_error_* api, but in this case there aren't
going to be any errors you can do anything about, so its safe to just
log the string.
```

I'm assuming from that response the the former will log the same error string as the second version.

Comment (korydraughn) [reply DSI/globus_gridftp_server_iRODS.cpp]: Sounds good to me.

**Comment by korydraughn** [comment]
If the tests pass, let's squash it.

**Comment by JustinKyleJames** [comment]
> If the tests pass, let's squash it.

done

**Review by korydraughn — APPROVED** [review]
Please pound if ready.

**Comment by JustinKyleJames** [comment]
> Please pound if ready.

done

**Review by alanking — APPROVED** [review]

**Pull request merged**
